misc-scripts/xref_mapping/XrefParser/MIMParser.pm

```diff
@@ -21,162 +21,245 @@ package XrefParser::MIMParser;
-			    desc       => $long_desc,
-			    source_id  => $morbid_source_id,
-			    species_id => $species_id,
-                            dbi        => $dbi,
-			    info_type  => "DEPENDENT"} );
-	}
-	elsif($type eq "^"){
-	  if(/\*FIELD\*\sTI\n[\^]\d+ MOVED TO (\d+)/){
-            if ($1 eq $number) { next; }
-	    $old_to_new{$number} = $1;
-	  }
-	  else{
-	    $removed{$number} = 1;
-	    $removed_count++;
-	  }
-	
-	}
+ RECORD:
+  while ( my $input_record = $mim_io->getline() ) {
+
+    my $ti = extract_ti( $input_record );
+    if ( ! defined $ti ) {
+      croak 'Failed to extract TI field from record';
+    }
+
+    my ( $type, $number, $long_desc ) = parse_ti( $ti );
+    if ( !defined( $type ) ) {
+      croak 'Failed to extract record type and description from TI field';
+    }
+
+    # Use the first block of text as description
+    my @fields = split( qr{;;}msx, $long_desc );
+    my $label = $fields[0] . " [" . $type . $number . "]";
+
+    my $xref_object = {
+                       acc        => $number,
+                       label      => $label,
+                       desc       => $long_desc,
+                       species_id => $species_id,
+                       dbi        => $dbi,
+                       info_type  => "DEPENDENT",
+                     };
+
+    if ( exists $TYPE_SINGLE_SOURCES{$type} ) {
+      my $type_source = $TYPE_SINGLE_SOURCES{$type};
+
+      $xref_object->{'source_id'} = $type_source;
+      $counters{ $type_source }++;
+      $self->add_xref($xref_object);
+
+    }
+    elsif ( $type eq q{+} ) {    # both gene and phenotype
+
+      $xref_object->{'source_id'} = $gene_source_id;
+      $counters{ $gene_source_id }++;
+      $self->add_xref($xref_object);
+
+      $xref_object->{'source_id'} = $morbid_source_id;
+      $counters{ $morbid_source_id }++;
+      $self->add_xref($xref_object);
+
+    }
+    elsif ( $type eq q{^} ) {
+      my ( $new_number ) = ( $long_desc =~ m{
+                                              MOVED\sTO\s
+                                              (\d+)
+                                          }msx );
+      if ( defined $new_number ) {
+        if ( $new_number ne $number ) {
+          $old_to_new{$number} = $new_number;
+        }
+      }
+      # Both leading and trailing whitespace has been removed
+      # so don't bother with another regex match, just compare.
+      elsif ( $long_desc eq 'REMOVED FROM DATABASE' ) {
+        $removed{$number} = 1;
+        $counters{ 'removed' }++;
       }
+      else {
+        croak "Unsupported type of a '^' record: '${long_desc}'\n";
+      }
+
     }
-  }
+
+  } ## record loop
 
   $mim_io->close();
 
-  my $syn_count =0;
-  foreach my $mim (keys %old_to_new){
-    my $old= $mim;
-    my $new= $old_to_new{$old};
-    while(defined($old_to_new{$new})){
+  foreach my $mim ( keys %old_to_new ) {
+    my $old = $mim;
+    my $new = $old_to_new{$old};
+    while ( defined( $old_to_new{$new} ) ) {
```

Comment:
This seems like a strange sort of while loop

Reply:
Having had a very careful look at this, it seems to work what it is supposed to do - which is making all synonyms point to their respective final entries regardless of how many time an entry might have been moved. Not entirely optimal performance-wise but overall, I am impressed! I have just added extensive comments to that bit of code, hopefully all will be clear now.